### PR TITLE
save segment when FV failed

### PR DESCRIPTION
**Pull request by oscar-davids**
**What does this pull request do? Explain your changes. (required)**
For #[352](https://github.com/livepeer/internal-project-tracking/issues/352)

**Specific updates (required)**

- 
- 
- 

**How did you test each of these updates (required)**

**Does this pull request close any open issues?**

**Checklist:**

- [x] Read the [contribution guide](./doc/contributing.md)
- [x] `make` runs successfully
- [x] All tests in `./test.sh` pass
- [ ] README and other documentation updated
- [ ] [Pending changelog](./CHANGELOG_PENDING.md) updated

### CODE REVIEW

File: server/broadcast.go [hunk 1]
```diff
@@ -559,12 +559,26 @@ func (bsm *BroadcastSessionsManager) chooseResults(ctx context.Context, submitRe
...
+
 			}
-			clog.Infof(ctx, "Video comparison from url=%s and url=%s are equal=%v",
+			clog.Infof(ctx, "Video comparison from url=%s and url=%s are equal=%v saveenable=%v",
 				trustedResult.TranscodeResult.Segments[segmToCheckIndex].Url,
-				untrustedResult.TranscodeResult.Segments[segmToCheckIndex].Url, vequal)
+				untrustedResult.TranscodeResult.Segments[segmToCheckIndex].Url, vequal, drivers.FailSaveEnabled())
+		} else {
+			if drivers.FailSaveEnabled() {
+				if drivers.FailSaveEnabled() {
```
Comment (cyberj0g) [inline server/broadcast.go]: Duplicate `if`?

File: server/broadcast.go [hunk 2]
```diff
@@ -559,12 +559,26 @@ func (bsm *BroadcastSessionsManager) chooseResults(ctx context.Context, submitRe
...
 				return nil, nil, err
 			}
-			if monitor.Enabled && !vequal {
-				monitor.FastVerificationFailed(ctx, ouri, monitor.FVType2Error)
+			if !vequal {
+				if monitor.Enabled {
+					monitor.FastVerificationFailed(ctx, ouri, monitor.FVType2Error)
+				}
+				if drivers.FailSaveEnabled() {
+					drivers.SavePairData2GS(trustedResult.TranscodeResult.Segments[segmToCheckIndex].Url, trustedSegm,
```
Comment (cyberj0g) [inline server/broadcast.go]: Should we upload async, since it's on the latency-critical path?

File: cmd/livepeer/livepeer.go [hunk 3]
```diff
@@ -167,6 +167,9 @@ func parseLivepeerConfig() starter.LivepeerConfig {
 	cfg.Objectstore = flag.String("objectStore", *cfg.Objectstore, "url of primary object store")
 	cfg.Recordstore = flag.String("recordStore", *cfg.Recordstore, "url of object store for recordings")
 
+	// Fast Verification GS bucket:
+	cfg.FVfailGsBucket = flag.String("FVfailGsbucket", *cfg.FVfailGsBucket, "Google storage bucket for storing segments for fast verification")
```
Comment (cyberj0g) [inline cmd/livepeer/livepeer.go]: Sorry for nitpicking, but let's make user-facing messages more concise, like:
```
Google Cloud Storage bucket for storing segments, which failed fast verification

Google Cloud Storage private key file name or key in JSON format for accessing FVfailGsBucket
```

**Review by cyberj0g — CHANGES_REQUESTED** [review]
The code is fine after addressing few comments. I have a slight concern regarding whether we should add this function to master branch, because users will never really need it, but it may be useful for an attacker to analyze verification behavior. Maybe we can just use this branch for testing? WDYT @thomshutt @yondonfu @oscar-davids

File: server/broadcast.go (hunk 2 above)
Comment (oscar-davids) [reply server/broadcast.go:567]: fixed.

File: cmd/livepeer/livepeer.go (hunk 3 above)
Comment (oscar-davids) [reply cmd/livepeer/livepeer.go]: fixed.

**Comment by oscar-davids** [comment]
> The code is fine after addressing few comments. I have a slight concern regarding whether we should add this function to master branch, because users will never really need it, but it may be useful for an attacker to analyze verification behavior. Maybe we can just use this branch for testing? 

It is currently used for analyzing Fast Verification fail cases but in the future, it would be used for Full Verification, I think.

File: server/broadcast.go (hunk 2 above)
Comment (thomshutt) [reply server/broadcast.go]: 👍 Definitely

**Comment by thomshutt** [comment]
> The code is fine after addressing few comments. I have a slight concern regarding whether we should add this function to master branch, because users will never really need it, but it may be useful for an attacker to analyze verification behavior. Maybe we can just use this branch for testing? WDYT @thomshutt @yondonfu @oscar-davids

Interesting point, would someone looking to attack it in that way not already have what they need from this code being open source though?

File: server/broadcast.go (hunk 2 above)
Comment (oscar-davids) [reply server/broadcast.go]: Fixed

**Comment by cyberj0g** [comment]
Right, the code is open source, but program arguments are lower hanging fruit for someone willing to do the research. Also, by adding parameters not directly required for performing intended function, we increasing potential attack surface in a broad sense (e.g. someone at some point could commit the code or post the comment with organization's GSC access credentials in the arguments, or there may be a critical vulnerability in GC library). However, if it's not a single-use addition, then I'm totally fine with merging it.

**Comment by thomshutt** [comment]
@cyberj0g fair points, I had the same initial reaction to merging this in too but I think it's going to be too difficult to diagnose any false FV failure in Production without it. I'd be open to removing it in the future once we're comfortable this is working as intended though.

@oscar-davids could you confirm that this won't currently break Production (i.e where we don't have the GC config values set) if we tried to deploy it?

**Comment by oscar-davids** [comment]
I am sure that this won't currently break the Production.

**Review by cyberj0g — APPROVED** [review]

**Pull request merged**